src/app/inventory/store/objectives.ts

```diff
@@ -61,3 +62,53 @@ export function isBooleanObjective(
       (!objectiveDef.allowOvercompletion || !objectiveDef.showValueOnComplete))
   );
 }
+
+export function isTrialsPassage(item: DimItem, defs: D2ManifestDefinitions) {
+  if (item.objectives?.length === 3 && item.isExotic) {
+    for (const objective of item.objectives) {
+      const objectiveDef = defs.Objective.get(objective.objectiveHash);
+      if (isFlawlessObjective(objective, objectiveDef)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+export function isFlawlessPassage(
+  objectives: DestinyObjectiveProgress[] | null,
+  defs: D2ManifestDefinitions
+) {
+  if (objectives && objectives.length === 3) {
+    for (const objective of objectives) {
+      if (isFlawlessObjective(objective, defs.Objective.get(objective.objectiveHash))) {
+        return objective.complete;
+      }
+    }
+    return false;
+  }
+  return false;
+}
+
+// Assumes that the item related to the objective is a trials passage
+export function isFlawlessObjective(
+  objective: DestinyObjectiveProgress,
+  objectiveDef: DestinyObjectiveDefinition
+) {
+  return (
+    objective.completionValue === 1 &&
+    objectiveDef.allowValueChangeWhenCompleted &&
+    !objectiveDef.allowOvercompletion
```

Comment:
Are we sure the flawless objective is the only one that could match this?

Reply:
There is no guarantee so I think the best way would be to create a json file with a list of the hashes I need using [https://github.com/DestinyItemManager/d2-additional-info](url) as @delphiactual had originally suggested. That way I can easily check for the hashes and not worry about the objective definitions. I'll make a PR there and then implement the logic in here once it's been approved.